### PR TITLE
Add constraint event_dim

### PR DESCRIPTION
**Pull request by fehiepsi**
Addresses #874.

In this PR, I copied @fritzo's effort at https://github.com/pytorch/pytorch/pull/50547 and added his suggestion for IndependentTransform. Also removing `input_event_dim/output_event_dim` and deprecating `Transform.event_dim`.

### CODE REVIEW

File: numpyro/distributions/distribution.py [hunk 1]
```diff
@@ -797,11 +798,13 @@ def rsample(self, key, sample_shape=()):
...
-            t.domain = domain
-            domain = t.codomain
-        return domain
+        codomain = self.transforms[-1].codomain
+        codomain_event_dim = codomain.event_dim
+        assert self.event_dim >= codomain_event_dim
+        if self.event_dim == codomain_event_dim:
+            return codomain
+        else:
+            return independent(codomain, self.event_dim - codomain_event_dim)
```
Comment (fritzo) [inline numpyro/distributions/distribution.py:807]: I believe this needs more complex logic as in https://github.com/pytorch/pytorch/pull/50581/files#diff-8b108de0fae81bfc4f939e40322885653fc3c957b2f8ad0ff25aafc16bf2c639

File: numpyro/distributions/transforms.py [hunk 2]
```diff
@@ -168,18 +152,28 @@ def codomain(self):
...
+    for part in parts[len(parts) - 1::-1]:
+        input_event_dim = part.domain.event_dim + max(input_event_dim - part.codomain.event_dim, 0)
+    return input_event_dim
+
+
+def _get_compose_transform_output_event_dim(parts):
+    output_event_dim = parts[0].codomain.event_dim
+    for part in parts[1:]:
+        output_event_dim = part.codomain.event_dim + max(output_event_dim - part.domain.event_dim, 0)
+    return output_event_dim
```
Comment (fritzo) [inline numpyro/distributions/transforms.py:176]: This is a bit trickier in case any parts change event dim. I believe there are correct implementations in https://github.com/pytorch/pytorch/pull/50581/files#diff-9c38272c807be20a5832ca58bc3b1f3454c70d30e28ecbd39ca40191219f7d81

**Review by fritzo — CHANGES_REQUESTED** [review]
@fehiepsi could you also take a look at the updated PR https://github.com/pytorch/pytorch/pull/50581?

File: numpyro/distributions/distribution.py (hunk 1 above)
Comment (fehiepsi) [reply numpyro/distributions/distribution.py:807]: Yes, I added a comment there. Here we fixed `self.event_dim` directly from the constructor.

File: numpyro/distributions/transforms.py (hunk 2 above)
Comment (fehiepsi) [reply numpyro/distributions/transforms.py:176]: I just saw that we are making duplicated effort (they are equivalent, I did comment about the reason in your PR). ;)

File: numpyro/distributions/distribution.py (hunk 1 above)
Comment (fritzo) [reply numpyro/distributions/distribution.py:807]: Ok, it seems reasonable to defer `.forward_shape()` to a separate PR.

**Review by fritzo — APPROVED** [review]

File: test/test_distributions.py [hunk 3]
```diff
@@ -230,81 +230,85 @@ def _is_batched_multivariate(jax_dist):
...
-    elif isinstance(constraint, constraints._IntegerGreaterThan):
+    elif isinstance(constraint, constraints.integer_greater_than):
         return constraint.lower_bound + random.poisson(key, np.array(5), shape=size)
-    elif isinstance(constraint, constraints._Interval):
+    elif isinstance(constraint, constraints.interval):
         lower_bound = jnp.broadcast_to(constraint.lower_bound, size)
         upper_bound = jnp.broadcast_to(constraint.upper_bound, size)
         return random.uniform(key, size, minval=lower_bound, maxval=upper_bound)
-    elif isinstance(constraint, (constraints._Real, constraints._RealVector)):
+    elif constraint in (constraints.real, constraints.real_vector):
```
Comment (neerajprad) [inline test/test_distributions.py:247]: FYI - we may expose these classes in the API (https://github.com/pytorch/pytorch/issues/50616) so that this `ininstance/is` distinction can be avoided.
